### PR TITLE
Fix load context bug

### PR DESCRIPTION
**Pull request by mfleader**
## Changes introduced with this PR

- [X] Fix bug 
- [x] Add unit test

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: loadfile/loadfile.go
Comment (dbutenhof) [inline loadfile/loadfile.go:30]: Calling `TrimPrefix` twice seems awkward, although it does allow working with or without a trailing separator. According to the documentation, `Abs` calls `Clean`, which will remove a trailing separator unless the full path is exactly `/`, so I think you can just trim `absDir + os.PathSeparator` unconditionally... although I suppose that's not a big deal unless this is called often enough that saving a few CPU cycles might be relevant. 😆

File: cmd/arcaflow/main.go
Comment (dbutenhof) [inline cmd/arcaflow/main.go:139]: Outside the scope of your change, really, but this error is a duplicate of line 122, and misleading if it ever happens since the failure is in loading the regular files from your context directory rather than `configFile`.

**Review by jaredoconnell — CHANGES_REQUESTED** [review]
Something broke:

```
joconnel@joconnel-mac arcaflow-engine % export WFPATH=/Users/joconnel/Documents/projects/arcaflow/workflows/test-workflows/python_deployer_wait_simple
joconnel@joconnel-mac arcaflow-engine % go run cmd/arcaflow/main.go -config=$WFPATH/config.yaml -input=$WFPATH/input.yaml -context=$WFPATH
2024-01-04T15:45:44-05:00	error	source=main	Failed to load configuration file /Users/joconnel/Documents/projects/arcaflow/workflows/test-workflows/python_deployer_wait_simple/config.yaml (error loading context files, failed to read file /Users/joconnel/Documents/projects/arcaflow-clean/arcaflow-engine (read /Users/joconnel/Documents/projects/arcaflow-clean/arcaflow-engine: is a directory))
```

We've allowed specifying the entire directory, which, from that, it determined the workflow path without it being explicitly set. That appears to be broken now.

**Comment by mfleader** [comment]
I initialized the slice to an incorrect length which added several empty strings that were converted to absolute file paths. It seems to be fixed now.

**Review by jaredoconnell — APPROVED** [review]
Looks good, and works.

File: loadfile/loadfile.go
Comment (webbnh) [inline loadfile/loadfile.go:20]: Why do you report `filepath.Base(filePath)` instead of the actual failing input, `filePath`?

File: loadfile/loadfile.go
Comment (webbnh) [inline loadfile/loadfile.go:21]: What lint complaint are you squelching here?

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:21]: I think you can omit the `_ = ` and just call `os.RemoveAll()` like a subroutine if you don't care about the value that it returns.

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:33]: It seems weird to intermingle the steps to create the file with those for creating the directory, given that they are independent of each other (so, swap lines 28 & 29).

It also seems weird to use `os.CreateTemp()`:  you don't need the random name (or, at least, the directory didn't need a random name, so it seems odd that the file would), so `os.Create()` seems like a better choice (it saves you from having to fetch the file name).

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:40]: Again, it seems weird to intermingle these steps.

Also, if you constuct the directory symlink path based on `dirpath` instead of `dirname`, I think you can avoid the call to `Join()`.

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:53]: Again, this is a strange ordering:  normally, you would test the result of the call immediately afterwards (e.g., because if it failed, then there is no point to constructing `filemapExp`; and, you can construct `filemapExp` at any point (e.g., you might do it at the beginning of the function for documentation purposes, or you might do it right before it is needed for context), so why do it here?

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:56]: I don't understand this comment:  based on the fact that `filemapExp` has _two_ entries, it seems like more than "only the regular file" will be loaded....

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:87]: File paths `"a"` and `"c"` are not materially different from each other, so I don't think there's much point in testing both of them.  Perhaps a value like `"rel/subdir/c.txt"` would be more useful.

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:93]: Why do you reference `testFilepaths` in lines 90 and 92 but not 91?  It seems like you should be consistent.

**Review by webbnh — COMMENTED** [review]
I've got a small question and a bunch of suggestions for polishing (but, I didn't see anything that I would block the merge over).

File: cmd/arcaflow/main.go
Comment (jaredoconnell) [reply cmd/arcaflow/main.go:45]: The linter will complain if we don't do that for exported (capital) properties, I believe.

File: loadfile/loadfile_test.go
Comment (jaredoconnell) [reply loadfile/loadfile_test.go:21]: Maybe. Sometimes the linter doesn't like that we're not using the value. This format explicitly shows that we're aware of the fact that we're throwing it away.

File: cmd/arcaflow/main.go
Comment (dbutenhof) [reply cmd/arcaflow/main.go:45]: Go is weird ... 😦

File: cmd/arcaflow/main.go
Comment (dbutenhof) [reply cmd/arcaflow/main.go:137]: The lack of encapsulation here also bothers me: that is, any failure in `ContextAbsFilepaths` is assumed explicitly to be "failed to determine absolute path of [...] context directory", which seems an unnecessary external dependency on the function's internal behavior. What we actually know here is that it failed, not really why. And since you're reporting the `err` value as part of the new `error`, a message along the lines that "determining absolute context file paths failed" would be sufficient.

File: loadfile/loadfile.go
Comment (dbutenhof) [inline loadfile/loadfile.go:10]: s/contents at each file/contents of each file/ ??

File: loadfile/loadfile.go
Comment (dbutenhof) [reply loadfile/loadfile.go:20]: This could result in awkward errors; for example I get that you don't want to report an error in `--workflow foo/bar.yaml` as `/var/tmp/whatever/foo/bar.yaml`, but I'm not sure that just `bar.yaml` is a useful improvement.

Then again, this should be a highly unlikely error as we already expanded any relative paths to `Abs` via the `Join`. (And in fact it would probably be better to detect and report *there* that the result of the `Join`, assuming that `Join` doesn't care, isn't a legal absolute path.)

File: loadfile/loadfile.go
Comment (dbutenhof) [inline loadfile/loadfile.go:23]: Seems unnecessarily indirect. I'd have gone with something more like "error loading context file %s (%w)". Although a better diagnosis (you don't currently retain enough information to do this easily, though) would be "error loading workflow file %s" (or config file, etc.: i.e., your `requiredFiles` keys.)

File: cmd/arcaflow/main.go
Comment (mfleader) [reply cmd/arcaflow/main.go:45]: the style linter gets its way

File: loadfile/loadfile.go
Comment (mfleader) [reply loadfile/loadfile.go:20]: yeah, I didn't want to report the whole filepath since it is probably quite long. I can change it to report the absolute filepath.

File: loadfile/loadfile.go
Comment (mfleader) [reply loadfile/loadfile.go:21]: From [gosec's readme](https://github.com/securego/gosec?tab=readme-ov-file#usage), `G304: File path provided as taint input`.

```shell
❯ golangci-lint run
loadfile/loadfile.go:21:20: G304: Potential file inclusion via variable (gosec)
		fileData, err := os.ReadFile(absPath)
```

I guess these are the same thing.

File: cmd/arcaflow/main.go
Comment (mfleader) [reply cmd/arcaflow/main.go:137]: The only error that can be thrown during `ContextAbsFilepaths` has to be raised from `filepath.Abs()`. On *nix systems, this leads to [unixAbs()](https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/path/filepath/path.go;drc=9e933c189ca3a84f12995b3c799364a06abc4376;l=297) which itself is only going to raise an error if something particularly weird happens during the call to [os.Getwd()](https://cs.opensource.google/go/go/+/master:src/os/getwd.go;drc=5755bad42adc23ad4a0c32149ac8cf78ece5d0b0;l=22), so it seemed safe to make the error message that specific.

File: loadfile/loadfile.go
Comment (mfleader) [reply loadfile/loadfile.go:21]: further investigation, gosec's [G304 page](https://securego.io/docs/rules/g304.html#the-right-way) recommends a clean, safe path. I added a clean path, but I don't know how to incorporate a "safe" path, or if the "safe" part is necessary since adding `filepath.Clean()` satisfied the linter.

File: cmd/arcaflow/main.go
Comment (mfleader) [reply cmd/arcaflow/main.go:137]: I believe to consistently open and read the intended files, that we want to use the absolute file paths to the config, input, and workflow, so I think we need to do this first.

```
A -> B := B depends upon A

determine absolute file path -> read config file -> unmarshal config yaml into config struct -> instantiate main logger
```

File: loadfile/loadfile_test.go
Comment (mfleader) [reply loadfile/loadfile_test.go:33]: I would still need to make some call somewhere to get the file's absolute path, either with `filepath.Join()` or `filepath.Abs()`.

File: cmd/arcaflow/main.go
Comment (webbnh) [reply cmd/arcaflow/main.go:45]: Oh, [bleh](https://google.github.io/styleguide/go/decisions#doc-comments).

At least they own the problem...sort of:
> This makes it easy to [blah-blah] it later by simply replacing the [blah] name with the newly-[blah] one **_across both comments and code_**.

(Emphasis mine.)

😝

File: loadfile/loadfile_test.go
Comment (webbnh) [reply loadfile/loadfile_test.go:21]: There are lots of calls made to functions which return values in which the returned value is not used.  Go itself does not object in these cases (I cannot comment upon the linter's behavior).

> This format explicitly shows that we're aware of the fact that we're throwing it away.

True.  The question is, is that valuable enough to justify the visual clutter.  For instance, apparently this call was "weird enough" to justify having a code comment attached...and yet the comment doesn't mention that the returned value is being discarded or why (although the "even if it's there" _hints_ at it).  It would be better if the comment were expanded and the assignment were omitted...if the linter will permit that.  And, if the linter won't, it would be good to understand _why_ the linter thinks that it is necessary to explicitly capture a value that we intend to ignore.

File: loadfile/loadfile.go
Comment (webbnh) [reply loadfile/loadfile.go]: > I didn't want to report the whole filepath since it is probably quite long.

The question is, does that length represent a problem?  On the other side of the trade-off is the fact that brevity might be omitting details which are necessary for the user to diagnose and resolve the problem.  Unless you're really sure that the user would know where to find the file, you really should offer the whole path -- the extra 50-60 characters to the message probably won't cause a problem anything like the one posed by not knowing which directory the file is in.

File: loadfile/loadfile.go
Comment (webbnh) [reply loadfile/loadfile.go]: Wow...that's an impressive level of paranoia...but, I suppose it's good.

However, I think you've found a bug in the linter, if all you needed to do was to add `Clean()` to the result returned by `Abs()`, because "[Abs calls Clean on the result](https://pkg.go.dev/path/filepath#Abs)".

(I concur that establishing "safety" is a bit of a challenge, so perhaps it is not necessarily required.)

Anyway, I originally asked the question because it's generally an anit-pattern to disable large chunks of the linter, especially without explanation.  (In fact, apparently there's a [lint check](https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/nolintlint/README.md) for that....)  So, it would have been better to disable just `G304` rather than all of `gosec` (which might have allowed me to answer my own question), but, of course, fixing the code instead of disabling the linter is generally an even better solution!

File: cmd/arcaflow/main.go
Comment (webbnh) [reply cmd/arcaflow/main.go]: Fair enough:  I missed the detail that opening the config file depended on having the absolute path.

Does the ordering between the call to `flag.Usage()` and the logger matter?  (Because it's reversed between the `Abs()` error and the `LoadContext()` error.)  It might be handy to keep them consistent.

File: loadfile/loadfile_test.go
Comment (webbnh) [reply loadfile/loadfile_test.go]: > I would still need to make some call somewhere to get the file's absolute path, either with filepath.Join() or filepath.Abs().

I see...so you're using `CreateTemp()` as a clever substitute for `Join()`....  I'm not sure that that's really a win:  given that creating the directory is done in four steps (assign `dirname`, construct `dirpath`, call `os.MkdirAll()`, and check the result), having the file creation done in an analogous four-step process (assign `filename`, construct the file path, call `os.Create()`, and check the result) is very easy for the reader to apprehend.

Nice touch on adding the `Close()` call, btw.  🙂

File: cmd/arcaflow/main.go
Comment (webbnh) [inline cmd/arcaflow/main.go:165]: Why did you invent `j` here?  Doesn't it have exactly the same values as the first item returned by `range`?  I.e., wouldn't this be better?
```go
	for j, f := range requiredFilesAbsPaths {
		requiredFilesAbsSlice[j] = f
	}
```

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:67]: You don't want to make any assertions about the contents of `err`?  (This would server to protect your users, as well as to get more than one bit of resolution on whether the CUT is functioning correctly.)

Ditto line 74.

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:83]: I think you're missing the word "which" before "joins" (and I would omit the comma...).

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:97]: 🤣
The comment is great (if largely superfluous), but I'm thinking that line 97 should look like this, given how lines 95 and 98 are constructed:
```go
		"b": testFilepaths["b"],
```

File: loadfile/loadfile_test.go
Comment (webbnh) [inline loadfile/loadfile_test.go:99]: An interesting thing to note is the result of `Join()` is apparently passed through `Clean()`, so line 98 is removing the `../`....  (It's kind of brave of you to put that component of the path at the beginning, where the test cannot really be sure of its effect -- I would have put it in the middle!)

File: cmd/arcaflow/main.go
Comment (webbnh) [reply cmd/arcaflow/main.go]: > I missed the detail that opening the config file depended on having the absolute path.

Ah!  It _wasn't_ depending on it until your recent update!  😈

**Review by webbnh — APPROVED** [review]
Looks good, although I have a few comments and pointed questions for your consideration before you merge.

File: cmd/arcaflow/main.go
Comment (mfleader) [reply cmd/arcaflow/main.go:165]: In your example, `j` is the key, and `f` is the value for the items in `requiredFilesAbsPaths`, whereas `requiredFilesAbsSlice` is an array that uses an integer index. Transforming a map's values into an array or list is still an [experimental feature](https://pkg.go.dev/golang.org/x/exp/maps#Values).

**Review by dbutenhof — APPROVED** [review]

**Review by dbutenhof — APPROVED** [review]

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
👍

**Review by webbnh — APPROVED** [review]
👍

File: loadfile/loadfile.go
Comment (webbnh) [reply loadfile/loadfile.go:21]: The call to `Clean()` is in a subroutine of `Join()` which is [here](https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/path/filepath/path_unix.go;drc=319b75ed339b9c66d6cc0519a9adc248526c67d7;l=45).

**Pull request merged**
